Adds a new halloween species: Spirits

JohnFulpWillard opened the pull request:
## About The Pull Request

Adds 2 new species: Spirits and Ghosts

Spirits are available roundstart during Halloween, Wabbajack and Xenobio's black slime extract
Ghosts are available through Magic mirrors

They fly around, and don't have legs and instead float around. They also can't get legs implanted onto themselves.

They also do have organs, so they are affected by flashbangs, they do get hungry, they do need oxygen to survive (they don't take damage in space but they do suffocate & get slowdown), and can process chems.
Gibbing a ghost gives ectoplasm, an ingredient for ghost burgers.

Chaplains also got a buff here, null rod-type weapons' bane is now against Spirit mobs, rather than hardcoded revenants. This means it now includes Spirits/Ghosts, but also Soulscythes & Cult shades.

Also re-adds https://github.com/tgstation/tgstation/pull/81630 which was reverted in https://github.com/tgstation/tgstation/pull/86506 which I assume was accidental.

### The difference between Spirits and Ghosts

Ghosts have an innate ability to become incorporeal, which allows them to phase through walls and stuff. Using this will immediately make them drop any non-ghost limb/organ (not implants cause I thought it would be funny). This ability is not available if they have holy water in their system, and like revenants they also can't walk over blessed tiles with it. They are also invisible to cameras while using this (not the obscura though).

Sprites taken from observers directly, if anyone wants to make custom sprites for them feel free. If anyone wants to make this obtainable somehow in-game as well I wouldn't be opposed, halloween is just where I thought it would fit most.

This also adds a lot of fixes that I encountered trying to add this, from systems that have been neglected throughout the years.

https://github.com/user-attachments/assets/e368d710-80a0-4c63-b271-1abe3dd41a5e

## Why It's Good For The Game

We haven't gotten a new halloween species in a long time and thought it would be fun if you can play as an actual ghost, the soul that remains after a person passes, so Halloween feels more haunted. It's overall made in just good fun, with a bonus that Ghosts are a cool species to play with as well for Wizards & maybe Chaplains in the future (Dead sect when?)

## Changelog

:cl:
add: Added a new halloween species: Spirits, a species without legs and instead floats.
add: Added a new magic mirror species: Ghosts, like spirits but with the ability to become incorporeal, traversing through solid wall.
fix: Mobs unable to use storage items now can't use storage items.
fix: Mobs unable to use items can now not open airlocks & closets
fix: Mobs unable to pick items up can no longer pick items up and immediately drop, moving one tile at a time.
fix: Mobs with intentional missing limbs (Alien larva) no longer show their limbs as missing on examine (again)
fix: Golems' pref page had a missing icon, it now has one.
/:cl:

JohnFulpWillard committed:
Adds a ghost species

JohnFulpWillard committed:
Create screenshot_humanoids__datum_species_ghost.png

JohnFulpWillard committed:
Update closets.dm

Jezzared commented:
Honestly, making cult shades use this instead could be VERY interesting.

JohnFulpWillard committed:
mobility garbage

JohnFulpWillard committed:
Update _eyes.dm

JohnFulpWillard commented:
> Honestly, making cult shades use this instead could be VERY interesting.

that'd be a pretty insane buff to cult tbh, while chaplain blessing tiles for cult cases is already something they benefit out of for wraiths, this species can at will be able to use their hands and they go around blocked areas through walls & space, it's just just too strong of an ability to give to a conversion antagonist. If we want this to be obtainable in-game I'd prefer it as some form of science thing limited to ~1 person per round, like a genetic meltdown sorta thing.

Tostah commented:
species with free infinite jaunt essentially. I kinda like the idea of this for the chaplain because they already have like armblades and whatever with the null rod

MrMelbert commented on code/modules/mob/living/carbon/human/species_types/ghost.dm:
```diff
@@ -0,0 +1,182 @@
+			TRAIT_PULL_BLOCKED,
+			TRAIT_UI_BLOCKED,
+			), SPECIES_TRAIT)
+		carbon_owner.mobility_flags = initial(carbon_owner.mobility_flags)
```
Pretty sure you shouldn't be messing with mobi flags directly

JohnFulpWillard replied:
oh oops forgot to remove it in https://github.com/tgstation/tgstation/pull/90711/commits/b632b5c272e0b0b443da6f51e1adefcbbd419afd

MrMelbert commented on code/modules/mob/living/carbon/human/species_types/ghost.dm:
```diff
@@ -0,0 +1,182 @@
+		if(!(organ.organ_flags & ORGAN_GHOST))
+			organ.Remove(owner, special = FALSE)
+
+	if(HAS_TRAIT(carbon_owner, TRAIT_NO_FLOATING_ANIM))
```
`HAS_TRAIT_FROM`, surely? Otherwise this glitches out if you're buckled to something, as an example

MrMelbert commented on code/modules/mob/living/carbon/human/species_types/ghost.dm:
```diff
@@ -0,0 +1,182 @@
+	var/datum/species/carbon_species = carbon_owner.dna.species
+
+	//drop any limbs & organs that can't phase, now that you're phasing.
+	for(var/obj/item/bodypart/bodypart as anything in carbon_owner.bodyparts)
+		if(!(bodypart.bodytype & BODYTYPE_GHOST))
+			bodypart.drop_limb(special = FALSE, dismembered = FALSE, move_to_floor = TRUE)
```
Obligatory "if a ghost's chest is bioscrambled / augmented does this gib them"

JohnFulpWillard replied:
It doesn't drop chests but I'll make the ability unavailable to be consistent.

MrMelbert commented on code/modules/mob/living/carbon/human/species_types/ghost.dm:
```diff
@@ -0,0 +1,182 @@
+	swap_mode()
+
+/datum/action/innate/toggle_passthrough/IsAvailable(feedback)
+	if(LAZYLEN(owner.reagents))
+		if(locate(/datum/reagent/water/holywater) in owner.reagents)
+			return FALSE
```
`has_reagent`

MrMelbert commented on code/game/atom/alternate_appearance.dm:
```diff
@@ -195,7 +195,7 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
 	if (istype(M, /mob/living/basic/construct/wraith))
 		return TRUE
-	if(isrevenant(M) || IS_WIZARD(M))
+	if(isrevenant(M) || IS_WIZARD(M) || isghostspecies(M))
```
Maybe this should just be a trait at this point

MrMelbert commented on code/datums/storage/storage.dm:
```diff
@@ -920,14 +920,18 @@ GLOBAL_LIST_EMPTY(cached_storage_typecaches)
 	if(!isliving(to_show) || !to_show.can_perform_action(parent, ALLOW_RESTING | FORBID_TELEKINESIS_REACH))
 		return FALSE
 
+	//because of the check above, it's safe to assume we're living now.
+	if(!(to_show.mobility_flags & MOBILITY_STORAGE))
+		return FALSE
```
Are open storage windows closed upon gaining `MOBILITY_STORAGE`?

MrMelbert reviewed (commented):
I already anticipate phasing being horrible to play against as security. Vampires lost their batform because it was only used to grief security and this seems pretty much the same.

The rest of this is very funny though 10/10

ElGitificador commented:
a ghost suffering from pressure and Cold?

ATH1909 commented:
> a ghost suffering from pressure and Cold?

I think the need for oxygen is a bit more jarring, imo

ATH1909 commented:
died 2021
born 2025
welcome back batform

Krail commented:
We lost runic golem jaunt from the pride mirror...
I guess this will have to do... beggars can't be choosers.

JohnFulpWillard committed:
Camera interactions

Livrah commented:
New species, ON TG STATION?? no way
They should be able to phase with Hauntium items.
Imo swap oxy dmg and pressure immune for some realism.

JohnFulpWillard committed:
HUDs are a trait and update accordingly

JohnFulpWillard committed:
Update _traits.dm

JohnFulpWillard committed:
Update _traits.dm

JohnFulpWillard committed:
Update _tongue.dm

JohnFulpWillard committed:
Fixes mistake in holy role huds

JohnFulpWillard committed:
Merge branch 'master' into ghost-species

JohnFulpWillard committed:
Merge branch 'master' into ghost-species

Loafin34 commented:
Hypothetically, since ghosts are affected by vaccuums, they shouldn't be immune to pressure.

JohnFulpWillard committed:
Ghost patch

Health doll updates when you change species, and also doesnt spam updates when you do either
Allows human species' subtype to properly lack legs, so adds a ghost human species subtype
Makes a bunch of tools (like health analyzer) get the limbs you should have, rather than using the global list of all limbs

JohnFulpWillard commented:
They aren't affected by space because it only affects "biological" bodyparts which I could add I don't really mind, they do take damage from fire and stuff though from my testing.

JohnFulpWillard committed:
Update screen_objects.dm

JohnFulpWillard committed:
Update helpers.dm

JohnFulpWillard committed:
this should fix screenshot tests

JohnFulpWillard committed:
Merge remote-tracking branch 'upstream/master' into ghost-species

JohnFulpWillard committed:
Update deaconize.dm

JohnFulpWillard commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -1071,6 +1071,10 @@
 
 	synchronize_bodytypes()
 	synchronize_bodyshapes()
+	// Bodyparts need to be sorted for leg masking to be done properly. It also will allow for some predictable
+	// behavior within said bodyparts list. We sort it here as we're adding the bodypart. This can't be in try_attach_limb,
+	// because we're also called in create_bodyparts, where updating health and such is something we wish to avoid.
+	bodyparts = sort_list(bodyparts, GLOBAL_PROC_REF(cmp_bodypart_by_body_part_asc))
 
```
```suggestion

```

JohnFulpWillard committed:
Update code/modules/mob/living/carbon/carbon.dm

JohnFulpWillard committed:
Merge branch 'master' into ghost-species

JohnFulpWillard committed:
Heretic can see blessed tiles

JohnFulpWillard committed:
Adds spirits

JohnFulpWillard committed:
tests

JohnFulpWillard committed:
Update ghost.dm

LemonInTheDark commented:
what's goin on here?

JohnFulpWillard committed:
spirits dont have tails

Time-Green commented on code/datums/storage/storage.dm:
```diff
@@ -922,14 +922,18 @@ GLOBAL_LIST_EMPTY(cached_storage_typecaches)
 
 /// Opens the storage to the mob, showing them the contents to their UI.
-/datum/storage/proc/open_storage(mob/to_show)
+/datum/storage/proc/open_storage(mob/living/to_show)
 	if(isobserver(to_show))
 		show_contents(to_show)
 		return FALSE
```
Imma be real this is psycho behaviour. I think it's fine but still

Time-Green commented on code/game/objects/effects/blessing.dm:
```diff
@@ -9,19 +9,18 @@
-		I.alpha = 64
-		I.appearance_flags = RESET_ALPHA
-		add_alt_appearance(/datum/atom_hud/alternate_appearance/basic/blessed_aware, "blessing", I)
+	var/image/I = image(icon = 'icons/effects/effects.dmi', icon_state = "blessed", layer = ABOVE_NORMAL_TURF_LAYER, loc = src)
+	I.alpha = 64
+	I.appearance_flags = RESET_ALPHA
+	add_alt_appearance(/datum/atom_hud/alternate_appearance/basic/blessed_aware, "blessing", I)
```
While you're here, could you change the single letter variable name?

Time-Green commented on code/modules/mob/living/carbon/human/species_types/ghost.dm:
```diff
@@ -0,0 +1,256 @@
+	var/datum/action/innate/toggle_passthrough/passthrough_ability
+
+/datum/species/spirit/ghost/get_physical_attributes()
+	return "Ghosts are the spiritual remains of long-passed entities. They lack legs, can fly, can choose at will to become opaque, \
```
```suggestion
	return "Ghosts are the spiritual remains of long-passed entities. They lack legs, can fly, can choose at will to become incorporeal, \
```
Opaque just refers to becoming non-transparent, unless that's what you're going for for the lore here

Time-Green reviewed (changes requested):
Looks good to merge! Small changes suggested, and you got some conflicts

JohnFulpWillard committed:
Update code/modules/mob/living/carbon/human/species_types/ghost.dm

Co-authored-by: Time-Green <7501474+Time-Green@users.noreply.github.com>

JohnFulpWillard committed:
Update blessing.dm

JohnFulpWillard committed:
Merge remote-tracking branch 'upstream/master' into ghost-species

JohnFulpWillard committed:
a

JohnFulpWillard committed:
eyes
